### PR TITLE
Feature/event logging

### PR DESCRIPTION
**Pull request by mtrupkin**

### CODE REVIEW

File: proctor/pom.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
 		<tds-dll.version>3.0.0</tds-dll.version>
 		<progman-client.version>3.0.0</progman-client.version>
 		<sb11-mna-client.version>0.0.4-SNAPSHOT</sb11-mna-client.version>
-		<sb11-shared-code.version>3.0.0</sb11-shared-code.version>
+		<sb11-shared-code.version>3.0.1-SNAPSHOT</sb11-shared-code.version>
```
Comment (wtritch) [inline proctor/pom.xml]: Please release sb11-shared-code version 3.0.1 and bump this to a released version. =)

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 2]
```diff
@@ -201,9 +203,17 @@ private void validatePassword (String userPassword, String rtsPassword) {
   public boolean doLogout () {
     try {
       ProctorUser thisUser = getThisUser ();
-      if (thisUser != null && thisUser.getBrowserKey ().compareTo (Constants.UUIDEmpty) != 0)
-        _proctorUserService.logout (thisUser.getKey (), thisUser.getBrowserKey ());
-      clearOutProctorCookieInformationOrRemove (true);
+      try {
+        ProctorEventLogger.eventEntry(LOGOUT, thisUser.getId(), thisUser.getSessionKey());
+        if (thisUser != null && thisUser.getBrowserKey().compareTo(Constants.UUIDEmpty) != 0)
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: You're de-referencing thisUser right before a thisUser != null check.  You may want to move the logging statement inside the if check.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 3]
```diff
@@ -201,9 +203,17 @@ private void validatePassword (String userPassword, String rtsPassword) {
...
+          _proctorUserService.logout(thisUser.getKey(), thisUser.getBrowserKey());
+        clearOutProctorCookieInformationOrRemove(true);
+        ProctorEventLogger.info(LOGOUT, thisUser.getId(), thisUser.getSessionKey());
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: thisUser may be null here as well

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 4]
```diff
@@ -201,9 +203,17 @@ private void validatePassword (String userPassword, String rtsPassword) {
...
+      } catch (Exception e) {
+        ProctorEventLogger.error(LOGOUT, thisUser.getId(), thisUser.getSessionKey(), e);
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: thisUser may be null here as well =)  I'll stop now.

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java [hunk 5]
```diff
@@ -201,9 +203,17 @@ private void validatePassword (String userPassword, String rtsPassword) {
...
-      if (thisUser != null && thisUser.getBrowserKey ().compareTo (Constants.UUIDEmpty) != 0)
-        _proctorUserService.logout (thisUser.getKey (), thisUser.getBrowserKey ());
-      clearOutProctorCookieInformationOrRemove (true);
+      try {
+        ProctorEventLogger.eventEntry(LOGOUT, thisUser.getId(), thisUser.getSessionKey());
+        if (thisUser != null && thisUser.getBrowserKey().compareTo(Constants.UUIDEmpty) != 0)
+          _proctorUserService.logout(thisUser.getKey(), thisUser.getBrowserKey());
+        clearOutProctorCookieInformationOrRemove(true);
+        ProctorEventLogger.info(LOGOUT, thisUser.getId(), thisUser.getSessionKey());
+      } catch (Exception e) {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: I know there isn't much to be done, since the existing code checks Exception as well... but I wish this was a little more narrowly-scoped.
nit: final

File: proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java (hunk 4 above)
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Presentation/LoginPresenter.java]: Why not move this logging into the outer `catch (Exception e) {` block?

File: proctor/src/main/java/TDS/Proctor/Services/remote/RemoteTestOpportunityService.java [hunk 6]
```diff
@@ -226,7 +226,7 @@ public boolean pauseOpportunity(final UUID oppKey, final UUID sessionKey, final
         return testOpportunityService.pauseOpportunity(oppKey, sessionKey, proctorKey, browserKey);
     }
 
-    private static Map<Integer, Set<String>> parseAccommodations(final String accommodationsString) {
+    public static Map<Integer, Set<String>> parseAccommodations(final String accommodationsString) {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Services/remote/RemoteTestOpportunityService.java]: Why does this need to be public?

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 7]
```diff
@@ -200,7 +230,14 @@ private ReturnStatus proctorPing (@RequestParam (value = "sessionKey", required
...
       ProctorUser thisUser = checkAuthenticatedAndValidate(sessionKey, "ProctorPing");
 
-      _proctorAppTasks.getTestSessionTasks ().setSessionDateVisited (sessionKey, thisUser.getKey (), thisUser.getBrowserKey ());
+      ProctorEventLogger.eventEntry(PING, thisUser.getId());
+      try {
+        _proctorAppTasks.getTestSessionTasks().setSessionDateVisited(sessionKey, thisUser.getKey(), thisUser.getBrowserKey());
+        ProctorEventLogger.eventPing(thisUser.getId(), thisUser.getSessionKey());
+      } catch (Exception e) {
+        ProctorEventLogger.error(PING, thisUser.getId(), e);
+        throw e;
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Recommend moving this to the outer `catch (Exception e)` block rather than creating a nested try/catch.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 8]
```diff
@@ -240,28 +277,74 @@ public SessionDTO autoRefreshData (@RequestParam (value = "sessionKey", required
       UUID sessionKey = UUID.fromString(strSessionKey);
       ProctorUser thisUser = checkAuthenticatedAndValidate(sessionKey, "AutoRefreshData");
 
+      _logger.info("session-id: {} ", sessionKey);
+      _logger.info("browser-id: {} ", thisUser.getBrowserKey());
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: As long as we're in here, can we rename `_logger` to `logger` or `LOG` ? =)

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 9]
```diff
@@ -240,28 +277,74 @@ public SessionDTO autoRefreshData (@RequestParam (value = "sessionKey", required
...
+
       boolean bGetCurTestees = true; // always get current testees if parameter
-                                     // not exists
-      if (!StringUtils.isEmpty (strBGetCurTestees)) {
-        bGetCurTestees = Boolean.parseBoolean (strBGetCurTestees);
+      // not exists
+      if (!StringUtils.isEmpty(strBGetCurTestees)) {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Could change this to a positive `StringUtils.isNotEmpty(strBGetCurTestees)` check  ...but cleaning up the existing code is probably out of scope.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 10]
```diff
@@ -438,9 +536,8 @@ public SessionDTO insertSessionTests (@RequestParam (value = "sessionKey", requi
   @ResponseBody
   public SessionDTO pauseSession (@RequestParam (value = "sessionKey", required = false) String strSessionKey) throws ReturnStatusException, TDSSecurityException {
     ProctorUser thisUser = checkAuthenticated();
-    try {
       UUID sessionKey = UUID.fromString (strSessionKey);
-
+    try {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Interesting, why move the UUID parsing out of the try block?

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 11]
```diff
@@ -478,19 +575,29 @@ public TesteeRequestDTO getCurrentRequests (@RequestParam (value = "oppKey", req
...
-      BrowserAction browserAction = browserValidation.Check(BrowserInfo.GetHttpCurrent());
-      
-      // convert dates to client's time zone
-      _proctorAppTasks.getRequestTasks ().convertDates (testeeRequests, getTimezoneOffset ());
-      TesteeRequestDTO testeeRequestsDTO = new TesteeRequestDTO ();
-      testeeRequestsDTO.setRequests (testeeRequests);
-      testeeRequestsDTO.setBrowserAction (browserAction);
+      ProctorEventLogger.eventEntry(PRINT, APPROVE, thisUser.getId(), thisUser.getSessionKey());
+      TesteeRequestDTO testeeRequestsDTO;
+      try {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Can we use the containing try/catch block rather than create a new nested one?

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java [hunk 12]
```diff
@@ -514,8 +621,14 @@ public ReturnStatus denyTesteeRequest (@RequestParam (value = "requestKey", requ
     try {
       ProctorUser thisUser = checkAuthenticated();
       UUID requestKey = UUID.fromString (strRequestKey);
-
-      _proctorAppTasks.getRequestTasks ().denyTesteeRequest (thisUser.getSessionKey (), thisUser.getKey (), thisUser.getBrowserKey (), requestKey, reason);
+      ProctorEventLogger.eventEntry(PRINT, DENY, thisUser.getId(), thisUser.getSessionKey());
+      try {
```
Comment (wtritch) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Likewise here, I'll stop asking after this, but if we're already in a try/catch (Exception e) block, why create another?

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    READ,
+    APPROVE,
+    DENY
+  }
+
+  public static final String PENDING_STATUS = "pending";
+  public static final String SEGMENT_ENTRY_STATUS = "segmentEntry";
+  public static final String SEGMENT_EXIT_STATUS = "segmentExit";
+
+  public static void eventAddAssessments(final String proctorId, final UUID sessionId, final String[] assessments) {
```
Comment (wtritch) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: Maybe some javadocs about what these represent?

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    info(APP, event, null, null, null, fields(proctorId));
+  }
+
+  public static void info(final ILogEvent event, final String proctorId, final UUID sessionId) {
+    info(APP, event, null, sessionId.toString(), null, fields(proctorId, sessionId));
+  }
+
+  public static void info(final ILogEvent event, final SubEvent subEvent, final String proctorId, final UUID
+    sessionId) {
+    info(APP, event, null, sessionId.toString(), null, fields(proctorId, sessionId, subEvent));
```
Comment (wtritch) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: One of the downsides of public static methods is that mocking/testing them is difficult.  You may want to consider registering an injecting an EventLogger instance rather than using public static methods so your tests can validate that the expected information is being logged?

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java (hunk 8 above)
Comment (millerts) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Though I agree the naming change is preferred I believe Mike was trying to be as light as possible when touching the existing code.  I understand it is just a rename so it shouldn't cause any issues, but the change set for this file would grow a bit.

File: proctor/src/main/java/TDS/Proctor/Services/remote/RemoteTestOpportunityService.java (hunk 6 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Services/remote/RemoteTestOpportunityService.java]: The parsed accommodations were being logged.  If we wanted the only the raw accommodations string passed from the browser to the server, this call would not need to be made.  We could also not log the accommodations.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java (hunk 10 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: It looks like an unnecessary change.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java (hunk 11 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: I would like to.  We wouldn't be able to log the proctor id and session key for the exception.  I was aiming for being able to associate the exception with the that information in the logging trail.

File: proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java (hunk 12 above)
Comment (mtrupkin) [inline proctor/src/main/java/TDS/Proctor/Web/Handlers/ActiveSessionXHR.java]: Its the same situation here.

File: proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java (hunk 14 above)
Comment (mtrupkin) [inline proctor/src/main/java/org/opentestsystem/delivery/logging/ProctorEventLogger.java]: I will write tests and we feel good about changing the structure of the logger if needed.  I think this is the 4th version.  I appreciate your comments.

**Comment by mtrupkin** [comment]
Researching alternate approach to event logging in the feature/event-interceptor branch.
Looking to inspect the http request/response in order to determine the event type and event specific data.
